File: luaServices/mem.lua
types = {}
table.insert(types, "collector")
table.insert(types, "enhancer")

cfg = {}
cfg["name"] = "mem"
cfg["poolfreq"] = 60
cfg["history"]  = 25


function declare ()
	this.addRessource("mem", "Service memory usage", "serviceMemory")
	this.addProperty( "mem", "total",   "Total Memory usage (kB)", "number")
	this.addProperty( "mem", "swap",    "Swapped (kB)", "number")
	this.addProperty( "mem", "shared",  "Shared Libraries (kB)", "number")
end

function collect ()
	local f
	local i=1
	this.nextValue("mem")
	local total  = 0
	local shared = 0
	local swap   = 0
	local key    = ""
	this.getPIDList()
	while (type(pids[i]) == "number")
	do
		f = io.open("/proc/".. pids[i] .."/smaps", "r")
		if f~=nil then
			for line in f:lines() do
				key = line:match("(%a+)")
				if (key == "Rss") then
					val = line:match("(%d+)")
					total = total + tonumber(val)
				elseif (key == "Swap") then
					val = line:match("(%d+)")
					swap = swap + tonumber(val)
				elseif (key == "Shared_Clean") then
					val = line:match("(%d+)")
					shared = shared + tonumber(val)
				end
			end
			f:close()
		end
		i=i+1
	end
	this.setDProperty("mem", "swap",   swap)
	this.setDProperty("mem", "total",  total)
	this.setDProperty("mem", "shared", shared)
end

function enhance(serv)
	-- adding this collector to every service
	--io.write(string.format("Adding cpu collector to a service\n"))
	serv.addCollector("mem")
end

File: services/mem.cpp
#include "services.h"
#include <fstream>
#include <iostream>
#include <chrono>

// This is exactly the same as in the mem.lua file. This is here for memory usage difference only
using namespace watcheD;
namespace services {

class ServMemCollector : public Collector {
public:
	ServMemCollector(std::shared_ptr<HttpServer> p_srv, Json::Value* p_cfg, std::shared_ptr<service> p_serv): Collector("mem", p_srv, p_cfg, 25, 60, p_serv) {
		addRessource("mem", "Service memory usage", "serviceMemory");
		ressources["mem"]->addProperty("total",  "Total Memory usage (kB)",	"number");
		ressources["mem"]->addProperty("swap",   "Swapped (kB)",		"number");
		ressources["mem"]->addProperty("shared", "Shared Libraries (kB)",	"number");
		addGetMetricRoute();
		morrisType="Area";morrisOpts="  ";
	}

	void collect() {
		std::shared_ptr<service> serv_p = onService.lock();
		if (!serv_p) return;
		std::shared_ptr< std::vector<uint32_t> > pids = serv_p->getPIDs();
		uint32_t total  = 0;
		uint32_t shared = 0;
		uint32_t swap   = 0;
		for (std::vector<uint32_t>::iterator i=pids->begin();i!=pids->end();i++) {
			std::ifstream	infile("/proc/"+std::to_string(*i)+"/smaps");
			std::string	line;
			if (!infile.good()) continue;
			while(std::getline(infile, line)) {
				if (line.find(" kB") == std::string::npos) continue;
				std::string key   = line.substr(0,line.find(":"));
				std::string value = line.substr(key.length()+1, line.find(" kB"));
				if (key == "Rss")		total  += atoi(value.c_str());
				else if (key == "Swap")		swap   += atoi(value.c_str());
				else if (key == "Shared_Clean")	shared += atoi(value.c_str());
			}
			infile.close();
		}
		ressources["mem"]->nextValue();
		ressources["mem"]->setProperty("total",		total);
		ressources["mem"]->setProperty("swap",		swap);
		ressources["mem"]->setProperty("shared",	shared);
	}
};

MAKE_PLUGIN_SERVICE_COLLECTOR(ServMemCollector, mem)


}
